zvariant/src/lib.rs

```diff
@@ -1293,6 +1293,158 @@ mod tests {
+            ("123", ObjectPath::try_from("/hello").unwrap()),
+            ("456", ObjectPath::try_from("/world").unwrap()),
+        ]),
+        dict_iter_24: (&str, Signature<'_>, vec![
+            ("123", Signature::try_from("i").unwrap()),
+            ("456", Signature::try_from("u").unwrap()),
+        ]),
+    }
+
+    #[cfg(unix)]
+    dict_iter_tests_gen! {
+        dict_iter_25: (Fd<'_>, &str, vec![(mock_fd(0), "123"), (mock_fd(1), "456")]),
+        dict_iter_26: (&str, Fd<'_>, vec![("123", mock_fd(0)), ("456", mock_fd(1))]),
+    }
+
+    macro_rules! dict_iter_mut_tests_gen {
+        ($($name:ident, p2s: $k_ty:ty,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let inputs = vec![(1 as $k_ty, "123"), (2 as $k_ty, "456")];
+                    let expect = vec![(1 as $k_ty, "123-hello"), (2 as $k_ty, "456-hello")];
+                    let mut dict1 = Dict::new(<$k_ty>::signature(), <&str>::signature());
+                    for (k, v) in inputs {
+                        dict1.add(k, v).unwrap();
+                    }
+                    let mut dict2 = Dict::new(<$k_ty>::signature(), <&str>::signature());
+                    for (k, v) in expect {
+                        dict2.add(k, v).unwrap();
+                    }
+                    for (_, v) in dict1.iter_mut() {
+                        if let Value::Str(vv) = v {
+                            *vv = Str::from(vv.to_string() + "-hello");
+                        }
+                    }
+                    assert_eq!(dict1, dict2);
+                }
+            )*
+        };
+        ($($name:ident, s2p: ($k_ty:ty, $zv_ty:ident),)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let inputs = vec![("123", 1 as $k_ty), ("456", 2 as $k_ty)];
+                    let expect = vec![("123", 2 as $k_ty), ("456", 3 as $k_ty)];
+                    let mut dict1 = Dict::new(<&str>::signature(), <$k_ty>::signature());
+                    for (k, v) in inputs {
+                        dict1.add(k, v).unwrap();
+                    }
+                    let mut dict2 = Dict::new(<&str>::signature(), <$k_ty>::signature());
+                    for (k, v) in expect {
+                        dict2.add(k, v).unwrap();
+                    }
+                    for (_, v) in dict1.iter_mut() {
+                        if let Value::$zv_ty(vv) = v {
+                            *vv += 1 as $k_ty;
+                        }
+                    }
+                    assert_eq!(dict1, dict2);
+                }
+            )*
+        };
+    }
+
+    dict_iter_mut_tests_gen! {
+        // basic -> str
+        dict_iter_mut_01, p2s: u8,
+        dict_iter_mut_02, p2s: u16,
+        dict_iter_mut_03, p2s: i16,
+        dict_iter_mut_04, p2s: u32,
+        dict_iter_mut_05, p2s: i32,
+        dict_iter_mut_06, p2s: u64,
+        dict_iter_mut_07, p2s: i64,
```

Comment:
I don't think we need tests for each type. Just the case of a key: randomly selected basic type and string, and then viceversa is enough and then we don't need the macros.

Reply:
Testing with a good coverage of these makes us possible to find the f64 issue as early as possible which is the reason I added all these test cases for compatibility if we might have future signature changes.
If you would still like to have only 1 for `basic -> string` and 1 for `string -> basic`, I can update the PR.

Reply:
> Testing with a good coverage of these makes us possible to find the f64 issue

Sure but that's a one time thing and try it once and then you know (we don't need to test for it all the time). You can even leave the test case for `f64` as commented out with a comment on top.

> If you would still like to have only 1 for `basic -> string` and 1 for `string -> basic`, I can update the PR.

Yes please.

Reply:
I've removed the macro and only use u8 and f64 as `zvariant::Basic` test sample.
The code is available in the 3rd commit (adamcavendish/zbus@c2475a549f1f05d2b54f0de1cde5aa4a9b9d2da9)
Note: I'll squash the commits into one after all reviews are finished and turn the PR into ready.

Reply:
> Note: I'll squash the commits into one after all reviews are finished and turn the PR into ready.

Please always squash the commits and keep them atomic (in terms of  "one logical commit per commit" rather than "each development step in a separate commit). That helps with the review since then I know which commits with what log, I'm approving. It also makes it easy to do per-commit review.

Github provides a facility to compare the diff with the previous version of the commit(s), in case the reviewer is interested in that.